src/wfs/src/main/java/org/geoserver/wfs/outputformat/WFSOutputFormatCallback.java

```diff
@@ -0,0 +1,81 @@
+import org.geoserver.ows.AbstractDispatcherCallback;
+import org.geoserver.ows.Request;
+import org.geoserver.platform.Operation;
+import org.geoserver.platform.Service;
+import org.geoserver.platform.ServiceException;
+import org.geoserver.wfs.WFSInfo;
+import org.geotools.util.logging.Logging;
+
+/**
+ * A callback that will filter the {@link WFSGetFeatureOutputFormat} based on {@link
+ * WFSInfo#getGetFeatureOutputTypes()}
+ */
+public class WFSOutputFormatCallback extends AbstractDispatcherCallback {
+    private static final Logger LOGGER = Logging.getLogger(WFSOutputFormatCallback.class);
+    private final GeoServer geoserver;
+    private static final String WFS = "wfs";
+    private static final String GET_FEATURE = "GetFeature";
+
+    public WFSOutputFormatCallback(GeoServer geoserver) {
+        this.geoserver = geoserver;
+    }
+
+    /**
+     * Filters out Requests whose Output Format are not explicitly enabled if Get Feature Output
+     * Type Checking is enabled
+     *
+     * @param request The request.
+     * @param operation The operation for the request.
+     * @return operation if Requested Output Format is allowed
+     */
+    @Override
+    public Operation operationDispatched(Request request, Operation operation) {
+        Service service = operation.getService();
+        if (service == null
+                || service.getId() == null
+                || !service.getId().equalsIgnoreCase(WFS)
+                || !request.getRequest().equalsIgnoreCase(GET_FEATURE)) {
+            // not a WFS service or not getFeature so we are not interested in it
+            return operation;
+        }
+        WFSInfo wfs = geoserver.getService(WFSInfo.class);
+        if (!wfs.isGetFeatureOutputTypeCheckingEnabled()) {
+            // output type checking is disabled, so we are not interested in it
+            return operation;
+        }
+        // filter the output formats
+        String outputFormat = request.getOutputFormat();
+        if (outputFormat != null
+                && wfs.getGetFeatureOutputTypes() != null
+                && !wfs.getGetFeatureOutputTypes().contains(outputFormat)) {
+            LOGGER.fine(
+                    "Output Format "
+                            + outputFormat
+                            + " is not enabled for GetFeature due to Global WFS settings");
+            throw unallowedGetFeatureFormatException(outputFormat);
+        }
+        return operation;
+    }
+    /**
+     * Builds a ServiceException for an unallowed output format
+     *
+     * @param outputFormat the output format requested
+     * @return the ServiceException
+     */
+    public ServiceException unallowedGetFeatureFormatException(String outputFormat) {
+        ServiceException e =
+                new ServiceException(
+                        "Getting feature using " + outputFormat + " is not allowed",
+                        "ForbiddenFormat");
+        e.setCode("ForbiddenFormat");
```

Comment:
Redundant? You just set the code above. 
Also, codes are enumerated in the standard, I believe you should not make a new one. 
Use ServiceException.INVALID_PARAMETER_VALUE instead, while indicating in the logs (that the admin will access) that the format is actually forbidden. A client should just use the format names enumerated in the capablities document.

Reply:
done